crates/provider/src/quorum/provider.rs

```diff
@@ -0,0 +1,272 @@
+                )
+            })
+            .collect();
+        let transport = QuorumTransport::builder()
+            .quorum(quorum)
+            .add_transports(transports)
+            .build();
+        let inner = RootProvider::new(RpcClient::new(transport, false));
+        Self { inner }
+    }
+
+    pub fn builder() -> QuorumProviderBuilder {
+        QuorumProviderBuilder::default()
+    }
+}
+
+#[derive(Default)]
+pub struct QuorumProviderBuilder {
+    quorum: Quorum,
+    transports: Vec<WeightedTransport>,
+}
+
+impl QuorumProviderBuilder {
+    pub fn quorum(mut self, quorum: Quorum) -> Self {
+        self.quorum = quorum;
+        self
+    }
+
+    pub fn urls(mut self, urls: Vec<Url>) -> Self {
+        self.transports.extend(urls.into_iter().map(|url| {
+            WeightedTransport::new(
+                alloy_transport_http::Http::new(url).into_box_transport(),
+            )
+        }));
+        self
+    }
+
+    pub fn add_mocked_transport(mut self, asserter: Asserter) -> Self {
+        self.transports.push(WeightedTransport::new(
+            MockTransport::new(asserter).into_box_transport(),
+        ));
+        self
+    }
+
+    pub fn transports(mut self, transports: Vec<WeightedTransport>) -> Self {
+        self.transports.extend(transports);
+        self
+    }
+
+    pub fn build(self) -> QuorumProvider {
+        let transport = QuorumTransport::builder()
+            .quorum(self.quorum)
+            .add_transports(self.transports)
+            .build();
+
+        let inner = RootProvider::new(RpcClient::new(transport, false));
+        QuorumProvider { inner }
+    }
+}
+
+impl Provider for QuorumProvider {
+    fn root(&self) -> &RootProvider<Ethereum> {
+        &self.inner
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        Quorum,
+        QuorumProvider,
+        quorum::transport::WeightedTransport,
+    };
+    use alloy_json_rpc::ErrorPayload;
+    use alloy_primitives::private::serde::Serialize;
+    use alloy_provider::Provider;
+    use alloy_transport::{
+        IntoBoxTransport,
+        mock::{
+            Asserter,
+            MockTransport,
+        },
+    };
+
+    #[tokio::test]
+    async fn test_quorum_provider_majority_success() {
+        const EXPECTED_BLOCK_NUMBER: u64 = 42u64;
+
+        let transports = vec![
+            WeightedTransport::new(
+                MockTransport::new(successful_asserter(&EXPECTED_BLOCK_NUMBER))
+                    .into_box_transport(),
+            ),
+            WeightedTransport::new(
+                MockTransport::new(successful_asserter(&EXPECTED_BLOCK_NUMBER))
+                    .into_box_transport(),
+            ),
+            WeightedTransport::new(
+                MockTransport::new(successful_asserter(&EXPECTED_BLOCK_NUMBER))
+                    .into_box_transport(),
+            ),
+        ];
+
+        let provider = QuorumProvider::builder()
+            .quorum(Quorum::Majority)
```

Comment:
It also would be nice to have a test with `Majority` and 2 our of 3 are good, and 1 is bad response or different response(maybe we could have a test with error, and test with different result).